include/stl2/view/subrange.hpp

```diff
@@ -0,0 +1,237 @@
+			: subrange{r.begin(), r.end(), n} {}
+
+			template <ConvertibleTo<I> X, ConvertibleTo<S> Y>
+			constexpr subrange(pair<X, Y> r)
+			: subrange{r.first, r.second} {}
+
+			template <ConvertibleTo<I> X, ConvertibleTo<S> Y>
+			constexpr subrange(pair<X, Y> r, difference_type_t<I> n)
+			requires detail::True<Sized> // strange GCC bug, not yet reported
+			: subrange{r.first, r.second, n} {}
+
+			template <Range R>
+			requires ConvertibleTo<iterator_t<R>, I> && ConvertibleTo<sentinel_t<R>, S>
+			constexpr subrange(R& r)
+			: subrange{__stl2::begin(r), __stl2::end(r)} {}
+
+			template <Constructible<const I&> X, Constructible<const S&> Y>
+			requires ConvertibleTo<const I&, X> && ConvertibleTo<const S&, Y>
+			constexpr operator pair<X, Y>() const {
+				return pair<X, Y>{this->begin(), this->end()};
+			}
+
+			constexpr bool empty() const {
+				return this->begin() == this->end();
+			}
+			constexpr explicit operator bool() const {
+				return !empty();
+			}
+			constexpr bool operator!() const {
+				return empty();
+			}
+			constexpr auto size() const requires Sized {
+				return this->end() - this->begin();
+			}
+		};
+
+		template <Iterator I, Sentinel<I> S>
+		requires !SizedSentinel<S, I>
+		struct subrange<I, S, true>
+		: view_interface<subrange<I, S, true>>
+		{
+		private:
+			subrange<I, S> rng_ {};
+			difference_type_t<I> size_ = 0;
+		public:
+			using iterator = I;
+			using sentinel = S;
+
+			subrange() = default;
+
+			constexpr subrange(I i, S s, difference_type_t<I> n)
+			: rng_{i, s}, size_{n} {}
+
+			template <ConvertibleTo<I> X, ConvertibleTo<S> Y>
+			constexpr subrange(subrange<X, Y, true> r)
+			: subrange{r.begin(), r.end(), r.size()} {}
+
+			template <ConvertibleTo<I> X, ConvertibleTo<S> Y, bool B>
+			constexpr subrange(subrange<X, Y, B> r, difference_type_t<I> n)
+			: subrange{r.begin(), r.end(), n} {
+				if constexpr (B) {
+					STL2_EXPECT(n == r.size());
+				}
+			}
+
+			template <ConvertibleTo<I> X, ConvertibleTo<S> Y>
+			constexpr subrange(pair<X, Y> r, difference_type_t<I> n)
+			: subrange{r.first, r.second, n} {}
+
+			template <SizedRange R>
+			requires Constructible<subrange<I, S>, R&>
+			constexpr subrange(R& r)
+			: subrange{__stl2::begin(r), __stl2::end(r), __stl2::size(r)} {}
+
+			template <Constructible<const I&> X, Constructible<const S&> Y>
+			requires ConvertibleTo<const I&, X> && ConvertibleTo<const S&, Y>
+			constexpr operator pair<X, Y>() const {
+				return pair<X, Y>{rng_.begin(), rng_.end()};
+			}
+
+			template <Constructible<const I&> X, Constructible<const S&> Y, bool B>
+			requires ConvertibleTo<const I&, X> && ConvertibleTo<const S&, Y>
+			constexpr operator subrange<X, Y, B>() const {
+				if constexpr (B) {
+					return {rng_, size_};
+				} else {
+					return rng_;
+				}
+			}
+
+			constexpr operator subrange<I, S> const &() const & noexcept {
+				return rng_;
+			}
+
+			constexpr I begin() const {
```

Comment:
I find it odd that these return by value vs. reference for the other specialization.

Reply:
It is to defend the class invariant. If these return references, it's too easy to increment the begin iterator without decrementing the size.

I could make it return a const reference.

Reply:
Why is it more important to protect the class invariant for this specialization than the other specialization?

Reply:
Which invariant is that?

Reply:
That `begin()` and `end()` denote a range.

Reply:
The principle of least surprise. Everybody knows that you can't increment an iterator past the end. But they don't know that incrementing an iterator can violate that `subrange`'s invariant it it happens to be storing the size. I feel it's too irresponsible to be leaving that big of a foot gun lying around.

Reply:
That can never happen in generic `View`-handling code (for either case) since `ranges::begin` and `ranges::end` return prvalues. To violate the invariant, you must (a) know the concrete type of your range is an instance of `subrange`, and (b) intentionally modify `r.begin()` or `r.end()` in a way that would violate the invariant. I'm weakly inclined to allow that usage: code that intentionally mutates the `subrange` other than by assignment is opting into unsafe behavior.

I'm more strongly inclined to have consistent semantics for both specializations of `subrange`. I don't want code to mysteriously break a `forward_list` changes into a `list` or `vector`.

Reply:
Consider:

```c++
template <Range R>
void foo(R&& r)
{
  subrange sub{r};
  ++sub.begin();
  auto d = distance(sub);
  // ...
}
```

If `R` does not satisfy `SizedRange`, all is well. But if it _does_, you have a bug. That's way too subtle for me.

Reply:
My preference at the moment is to add `next`, `prev`, and `advance` to both specializations of `subrange` in a way that preserves the invariant in the `Sized` case.

Reply:
> If `R` does not satisfy `SizedRange`, all is well.

No: If `r` is empty, `sub` is invalid. I understand that there are more ways for a user to invalidate a `subrange` that is sized than a `subrange` that is not sized, but qualitatively they are similar. I think that this distinction between kinds of `subrange` will create confusion. I would rather we either (a) allow mutation for both kinds of `subrange`, consistently placing responsibility on the user, or (b) forbid mutation for both kinds of `subrange`, providing consistent safety.

Reply:
Then my preference is (b). I'm fine with that. I can add `next`, `prev`, and `advance` as the safe way to mutate the beginning of the sub-range.